### PR TITLE
Feature enumerables methods

### PR DESCRIPTION
**Pull request by martinnajjar12**
### Clone Enumerables Methods
We've used yield and call with Proc to create the following methods:

- #my_each
- #my_each_with_index
- #my_select
- #my_all?
- #my_any?
- #my_none?
- #my_count
- #my_map
- #my_inject

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,68 @@
...
+## Built With
+
+- Ruby
+
+## Getting Started
+
+To get a local copy up and running follow these simple example steps.
+
+- Clone this repository into your computer
+- To run, enter this command in your terminal (inside the project's directory) "ruby app.rb"
```
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** You can format` inline code like this` 👈 by surrounding the desired code with the backquote( ` )

File: README.md [hunk 2]
```diff
@@ -1 +1,68 @@
...
+Only Ruby.
+
+### Setup
+
+Ruby and gem.
+
+### Install
+
+Give the methods inside the file "app.rb" a random array and type in your terminal (inside the directory of the project) "ruby app.rb"
+
```
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** Remember that not anyone who checks your project is a tech person, so you may want to be more specific with instructions such as:

  - clone this repo with
  ```
  git clone <URL of this repo>
  ```

File: README.md [hunk 3]
```diff
@@ -1 +1,68 @@
...
+### Usage
+
+Practicing to build the methods manually
+
```
Comment (kitpao) [inline README.md]: - [x] In this "Usage" section is where you put how to execute the files, please put here the instructions to run them.
  **Hint** Optional recommended instructions:
  - open `app.rb` file with your favorite editor
  - add test methods like : `<give some examples>`
  - In your terminal run `ruby app.rb`

File: README.md [hunk 4]
```diff
@@ -1 +1,68 @@
...
+### Run tests
+
+rubocop
+
```
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** Rubocop is a linter, it doesn't count as a test, you can take it out. **Hint:** don't worry, in a future project, you will be required to add tests to your methods with RSpec

File: .rubocop.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+Style/EachForSimpleLoop:
+  Enabled: false
+Style/AndOr:
+  Enabled: false
+Style/DefWithParentheses:
+  Enabled: false
+Style/FrozenStringLiteralComment:
+  EnforcedStyle: never
+Style/For:
+  Enabled: false
```
Comment (kitpao) [inline .rubocop.yml]: - [x] These lines don't appear in the original document, kindly use the original file that you can find [here](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml).
For this project, it is allowed to disable the CaseEquality cop inside your own code without modifying .rubocop.yml, as explained in the pathwright
image
  image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @martinnajjar12 and @len23 

🌟 Great job implementing the new Git Flow for this project, you did it nicely
📝 Kindly check the suggested changes, in the inline comments you will find general fixes about your repo. And in this section, you will find everything about the failing cases for your methods and some hints to fix them.

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

### Methods:
  - [x] Please notice that the original methods take into account what happens when a block is not given (They return an enumerator object). This applies to my_each, my_each_with_index, my_select, my_map **Hint:** you can check [this documentation for more about enumerator objects](https://apidock.com/ruby/Object/to_enum).
  
- **Tip:** You can see the complete [description of all the methods here](https://ruby-doc.org/core-2.6.5/Enumerable.html), please check it out as a guide if you feel lost with any of the methods

#### #my_each_with_index
  - [x] Please check and correct the difference between original and your methods:
  ```
    p [1,2,3].each_with_index{ |x| x }               # [1,2,3]
    p [1,2,3].my_each_with_index{ |x| x }             # 0..2
  ```
  Notice that the original returns the original array, and yours is returning the last thing it used. **Hint:** Don't forget to return something at the end
 - [x] Your method should also work with ranges. **Hint:** you can convert the range into an array since the beginning
  ```
  p (1..3).each_with_index{ |x| x }               # 1..3
  p (1..3).my_each_with_index{ |x| x }             # Error  
  ```

#### #my_all?
  - [x] Please check and correct the difference between original and your method:
  ```
    p [1,2,3].all?               # true
    p [1,2,3].my_all?            # Error
  ```
  Notice that the original returns True and yours is returning an Error. **Hint:** This method should work even if a block is not given, also observe the behavior of the case `p [false, true, true].all?` -> false

  - [x] Please notice that your methods should be able to take Classes, Regexp and Patterns as parameters, 
**Hint:** Investigate the "===" operator, it can be very useful for this
  ```
  p [1,2,3].all?(Integer)              # true
  p [1,2,3].my_all?(Integer)           # `my_all?': wrong number of arguments (given 1, expected 0) (ArgumentError)
  ```

#### #my_any?
  - [x] Exact same issues as **#my_all?** (What happens when no block is given? What happens when an argument is given?)

#### #my_none?
   - [x] Exact same issues as **#my_all?** (What happens when no block is given? What happens when an argument is given?)

#### #my_count
  - [x] Please check and correct the difference between original and your method:
  ```
    p (5...50).count               # 45
    p (5...50).my_count            # Error
  ```
  Notice that the original returns the number of items, and yours is returning an Error. **Hint:** This method should work even if a block is not given
  
  - [x] This method should be able to receive parameters and count the items that are the same, check the example:
  ```
  p [1,1,2,2,2,2].count(2)        # 4
  p [1,1,2,2,2,2].my_count(2)     # `my_count': wrong number of arguments (given 1, expected 0) (ArgumentError)
  ```

#### #my_inject
  - [x]  `[2,3,4].my_inject(:+)` should return 9 -> 2+3+4=9. Your code doesn't know what to do with the symbol, it gives an error
  - [x] Your method should also be able to receive initial values with the symbol, `[2,3].my_inject(20, :*)` should return 120 (20 * 2 * 3), but right now, it gives an error

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate if can leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😛 

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General

- [x] `my_each` 
- it calls the given block once for each element in self
##### For example
````ruby
      my_each_output = ' ' 
      my_each_output_two = ' ' 
      block = proc { |x| my_each_output += "Hello: in: #{x}\n" }
      block_two = proc { |x| my_each_output_two += "Hello: in: #{x}\n" }
      [1,2,32,32,3,21,1,23,12].each(&block_two)
      [1,2,32,32,3,21,1,23,12].my_each(&block)
       p my_each_output ==  my_each_output_two ` this should return true
````
- [x] `my_each_with_index`  

- it calls the given block once for each element in self
##### For example
````ruby
      my_each_output = ' ' 
      my_each_output_two = ' ' 
      block = proc { |y, x| my_each_output += "Hello: #{y}, in: #{x}\n" }
      block_two = proc { |y, x| my_each_output_two += "Hello: #{y}, in: #{x}\n" }
      [1,2,32,32,3,21,1,23,12].each_with_index(&block_two)
      [1,2,32,32,3,21,1,23,12].my_each_with_index(&block)
       p my_each_output ==  my_each_output_two ` this should return true
````



- [x] `my_any?`:
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(Integer)` should return `true`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('cat')` should return `true`

- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`
`p (5..10).my_inject(:*)` should return `151200`
- Combines all elements of enum by applying a binary operation, specified by a block:
 `p (1..5).my_inject(4) { |prod, n| prod * n }` should return 480


- [x] `my_all?`  
- when a class is passed as an argument returns true if all of the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return `true`

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,274 @@
+require 'pry'
+# rubocop:disable Style/For
+# rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/ModuleLength, Metrics/MethodLength
+module Enumerable
+  def my_each
```
Comment (elmejdki) [inline app.rb:7]: ## my_each:

please make sure that your function return value matched the original one.

```ruby
  range = Range.new(5, 50)
  block = proc { |num| num < 9 / 2 }
  
  range.my_each(&block) === range.each(&block) => false

```

File: app.rb [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        for value in self do
+          yield(value)
+        end
+      end
+    else
+      to_enum :my_each
+    end
+  end
+
+  def my_each_with_index
```
Comment (elmejdki) [inline app.rb:25]: ## my_each_with_index:

```ruby

  range = Range.new(5, 50)
  block = proc { |num| num < 9 / 2 }

  range.my_each_with_index(&block) == range.each_with_index(&block) => false

  my_each_with_index_output == each_with_index_output => false

```

File: app.rb [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      my_each do |element|
+        yield(element) && new_arr << element
+      end
+      new_arr
+    else
+      to_enum :my_select
+    end
+  end
+
+  def my_all?(argument = nil)
```
Comment (elmejdki) [inline app.rb]: ## my_all:

```ruby
  numbers = [1, 2i, 3.14]

  numbers.my_all?(Numeric) == numbers.all?(Numeric) => false
```

File: app.rb [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        if element.class != argument
+          boolean = false
+          break
+        end
+      end
+    end
+    boolean
+  end
+
+  def my_any?(argument = nil)
```
Comment (elmejdki) [inline app.rb]: ## my_any?:

```ruby
  array = [1, 32,6,53,78,3,6,8,3,45,,86]

  array.my_any?(Numeric) == array.any?(Numeric) => false
```

File: app.rb [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+                     proc1.call(element)
+                   end
+      end
+      new_arr
+    else
+      to_enum :my_map
+    end
+  end
+
+  def my_inject(initial = nil, argument = nil)
```
Comment (elmejdki) [inline app.rb:283]: ## my_inject:

this method throws the following error:  
`undefined local variable or method `length' for 5..50:Range`

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 
## Well done guys :clap: :100: :+1: you have made a great job, but you still have to do some minor changes :hammer: :wrench: to get your project approved :checkered_flag: 

I have left some comments below please check them.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat:

File: app.rb [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+      else
+        for value in self do
+          yield(value)
+        end
+      end
+    else
+      to_enum :my_each
+    end
+    self
+  end
```
Comment (maelfosso) [inline app.rb]:
```ruby
2.6.5 :010 > array
 => [4, 6, 6, 8, 5, 4, 6, 3, 4, 8, 8, 3, 3, 1, 7, 2, 2, 2, 3, 6, 3, 5, 3, 3, 1, 4, 2, 0, 6, 5, 3, 8, 1, 8, 3, 8, 1, 4, 7, 4, 7, 6, 0, 0, 6, 7, 4, 0, 6, 3, 3, 0, 5, 0, 5, 6, 4, 4, 0, 1, 0, 6, 1, 1, 7, 4, 5, 6, 2, 5, 0, 0, 4, 4, 2, 6, 3, 8, 1, 6, 0, 4, 6, 3, 2, 0, 2, 5, 5, 7, 8, 8, 6, 3, 6, 7, 8, 5, 7, 3] 
2.6.5 :011 > array.each.is_a?(Enumerator)
 => true 
2.6.5 :012 > array.my_each.is_a?(Enumerator)
 => false 
2.6.5 :013 >
```

File: app.rb [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+            end
+      for index in 0..ary.length - 1 do
+        yield(ary[index], index)
+      end
+      ary
+    else
+      to_enum :my_each_with_index
+    end
+    self
+  end
```
Comment (maelfosso) [inline app.rb]:
```ruby
2.6.5 :015 > array
 => [4, 6, 6, 8, 5, 4, 6, 3, 4, 8, 8, 3, 3, 1, 7, 2, 2, 2, 3, 6, 3, 5, 3, 3, 1, 4, 2, 0, 6, 5, 3, 8, 1, 8, 3, 8, 1, 4, 7, 4, 7, 6, 0, 0, 6, 7, 4, 0, 6, 3, 3, 0, 5, 0, 5, 6, 4, 4, 0, 1, 0, 6, 1, 1, 7, 4, 5, 6, 2, 5, 0, 0, 4, 4, 2, 6, 3, 8, 1, 6, 0, 4, 6, 3, 2, 0, 2, 5, 5, 7, 8, 8, 6, 3, 6, 7, 8, 5, 7, 3] 
2.6.5 :016 > array.my_each_with_index.is_a?(Enumerator)
 => false 
2.6.5 :017 > array.each_with_index.is_a?(Enumerator)
 => true 
2.6.5 :018 > 
```

File: app.rb [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+    elsif argument.class == Integer || argument.class == Float || argument.class == String
+      my_each do |element|
+        if element == argument
+          boolean = false
+          break
+        end
+      end
+    end
+    boolean
+  end
```
Comment (maelfosso) [inline app.rb]:
```ruby
2.6.5 :020 > false_array = [nil, false, nil, false]
2.6.5 :021 > false_array.none?
 => true 
2.6.5 :022 > false_array.my_none?
 => false 
2.6.5 :023 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes.
However, there are still some issues to solve.
Kindly, read the comments left below to know more about them

Once done submit a new code review request
Thanks

Mael FOSSO

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap: 

Well done job :+1: 
You can merge it now :smiley_cat: 

Good continuation :handshake: 

Mael FOSSO

**Comment by martinnajjar12** [comment]
> # Congratulations 👏
> Well done job 👍
> You can merge it now 😺
> 
> Good continuation 🤝
> 
> Mael FOSSO

### Thank you so much 😊

**Pull request merged**